README
Flutter Android Bindings
========================

See <https://github.com/drydart/flutter_android> for documentation.

Use <https://github.com/drydart/flutter_android/issues> for bug reports.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.8.0] - 2020-08-19

### Changed

- Removed the dependency on the [meta](https://pub.dev/packages/meta) package
- Removed the dependency on the [platform](https://pub.dev/packages/platform)
  package

### Added

- `android_content.ComponentName#writeToParcel()` method
- `android_content.ContentValues#writeToParcel()` method
- `android_graphics.Point#writeToParcel()` method
- `android_graphics.PointF#writeToParcel()` method
- `android_os.Parcel#writeArrayMap()` method
- `android_os.Parcelable` mixin

## [0.7.0] - 2020-08-19

### Changed

- Required Dart 2.8+
- Required the [typed_data](https://pub.dev/packages/typed_data) package

### Added

- `android_os.Parcel.obtain()` static method
- `android_os.Parcel#asByteData()` method
- `android_os.Parcel#asUint8List()` method
- `android_os.Parcel#writeBoolean()` method
- `android_os.Parcel#writeBooleanArray()` method
- `android_os.Parcel#writeBundle()` method
- `android_os.Parcel#writeByte()` method
- `android_os.Parcel#writeByteArray()` method
- `android_os.Parcel#writeCharArray()` method
- `android_os.Parcel#writeDouble()` method
- `android_os.Parcel#writeDoubleArray()` method
- `android_os.Parcel#writeFloat()` method
- `android_os.Parcel#writeFloatArray()` method
- `android_os.Parcel#writeInt()` method
- `android_os.Parcel#writeIntArray()` method
- `android_os.Parcel#writeList()` method
- `android_os.Parcel#writeLong()` method
- `android_os.Parcel#writeLongArray()` method
- `android_os.Parcel#writeMap()` method
- `android_os.Parcel#writeParcelable()` method
- `android_os.Parcel#writeString()` method
- `android_os.Parcel#writeStringArray()` method
- `android_os.Parcel#writeValue()` method

## [0.6.2] - 2020-08-18

### Changed

- Enhanced `android_content.Intent#startActivity()`

## [0.6.1] - 2020-08-18

### Fixed

- Build error with the tagged 0.6.0 release

## [0.6.0] - 2020-08-18

### Added

- `android_content.Intent#package` property
- `android_content.Intent#startActivity()` method

## [0.5.0] - 2020-08-01

### Changed

- Required Flutter 1.17+ and Dart 2.7+

## [0.4.0] - 2019-03-13

### Added

- `android_hardware.Sensor.*` constants
- `android_hardware.SensorManager.*` constants
- `android_location.Location.*` constants
- `android_os.*.*` constants

### Removed

- Constants annotated upstream with `@hide` or `@deprecated`

## [0.3.7] - 2019-03-12

### Added

- `android_bluetooth.Bluetooth{Adapter,Device,Headset,LeScanner}.*` constants

## [0.3.6] - 2019-02-28

No functional changes.

## [0.3.5] - 2019-02-15

### Fixed

- Bug in `android_hardware.SensorManager.registerListener()`

## [0.3.4] - 2019-02-15

### Added

- `android_location.Location` class
- `android_hardware.Sensor#subscribe()` method
- `android_hardware.SensorEvent` class
- `android_hardware.SensorManager.getDefaultSensor()` static method
- `android_hardware.SensorManager.registerListener()` static method

## [0.3.3] - 2019-02-15

### Added

- `android_location.Location` class stub
- `android_location.Location.distanceBetween()` static method

## [0.3.2] - 2019-02-15

### Added

- Declared Bluetooth permissions for the example app
- Declared body sensor permissions for the example app
- `android_hardware.Sensor` class stub
- `android_hardware.SensorEvent` class stub
- `android_hardware.SensorEventListener` class stub
- `android_hardware.SensorManager` class stub

## [0.3.1] - 2019-02-14

### Added

- `android_bluetooth.BluetoothAdapter` class stub
- `android_bluetooth.BluetoothDevice` class stub
- `android_bluetooth.BluetoothHeadset` class stub
- `android_bluetooth.BluetoothLeScanner` class stub
- `android_bluetooth.BluetoothManager` class stub

## [0.3.0] - 2019-02-14

### Changed

- Required Dart 2.1+

### Added

- Face detection demo in the example app
- `android_graphics.Bitmap.fromImage()` static method
- `android_graphics.Bitmap.fromAssetImage()` static method
- `android_graphics.Bitmap#getHeight()` method
- `android_graphics.Bitmap#getWidth()` method
- `android_graphics.Bitmap#toImage()` method
- `android_graphics.Point` class
- `android_graphics.PointF` class
- `android_media.Face` class
- `android_media.FaceDetector` class

## [0.2.5] - 2019-02-13

### Added

- `android_graphics` library
- `android_graphics.Bitmap` class

## [0.2.4] - 2019-01-28

### Changed

- Added sanity checks to `android_database.MatrixCursor`

## [0.2.3] - 2018-12-10

### Added

- `android_content.Context.getSharedPreferences()` method
- `android_content.SharedPreferences` class

## [0.2.2] - 2018-12-08

### Changed

- Made `android_database.Cursor` iterable

## [0.2.1] - 2018-12-06

### Changed

- Enhanced the example application and added a screenshot

## [0.2.0] - 2018-12-05

### Changed

- `android_database.Cursor#getBlob()` now returns a `Uint8List`, not a `ByteBuffer`

## [0.1.14] - 2018-12-04

### Added

- `android_content.ContentValues` class

## [0.1.13] - 2018-11-23

### Added

- `android_app.ActivityManager.isRunningInTestHarness` getter

## [0.1.12] - 2018-11-22

### Changed

- Corrected the return value for `android_database.Cursor#moveToNext()`

## [0.1.11] - 2018-11-21

### Added

- `android_database.MatrixCursor` class

### Changed

- Corrected return values for `android_database.Cursor#moveTo*()` methods

## [0.1.10] - 2018-11-21

### Added

- `android_database.DatabaseUtils` static methods

## [0.1.9] - 2018-11-21

### Added

- `android_app.Notification` class
- `android_app.NotificationAction` class

## [0.1.8] - 2018-11-20

### Added

- [meta](https://pub.dev/packages/meta) dependency

## [0.1.7] - 2018-11-15

### Added

- `android_database.Cursor#get()` method

### Changed

- `android_database.CursorIndexOutOfBoundsException` constructor

## [0.1.6] - 2018-11-15

### Added

- `android_database.Cursor` interface

lib/src/os/parcel.dart
/* This is free and unencumbered software released into the public domain. */

import 'dart:typed_data'
    show
        ByteData,
        Endian,
        Float32List,
        Float64List,
        Int16List,
        Int32List,
        Int64List,
        Uint8List;

import 'package:typed_data/typed_buffers.dart' show Uint8Buffer;

import 'bundle.dart' show Bundle;
import 'parcelable.dart' show Parcelable;

const int VAL_NULL = -1;
const int VAL_STRING = 0;
const int VAL_INTEGER = 1;
const int VAL_MAP = 2;
const int VAL_BUNDLE = 3;
const int VAL_PARCELABLE = 4;
const int VAL_SHORT = 5;
const int VAL_LONG = 6;
const int VAL_FLOAT = 7;
const int VAL_DOUBLE = 8;
const int VAL_BOOLEAN = 9;
const int VAL_CHARSEQUENCE = 10;
const int VAL_LIST = 11;
const int VAL_SPARSEARRAY = 12;
const int VAL_BYTEARRAY = 13;
const int VAL_STRINGARRAY = 14;
const int VAL_IBINDER = 15;
const int VAL_PARCELABLEARRAY = 16;
const int VAL_OBJECTARRAY = 17;
const int VAL_INTARRAY = 18;
const int VAL_LONGARRAY = 19;
const int VAL_BYTE = 20;
const int VAL_SERIALIZABLE = 21;
const int VAL_SPARSEBOOLEANARRAY = 22;
const int VAL_BOOLEANARRAY = 23;
const int VAL_CHARSEQUENCEARRAY = 24;
const int VAL_PERSISTABLEBUNDLE = 25;
const int VAL_SIZE = 26;
const int VAL_SIZEF = 27;
const int VAL_DOUBLEARRAY = 28;

/// Container for a message (data and object references) that can be sent
/// through an `IBinder`.
///
/// See: https://developer.android.com/reference/android/os/Parcel
class Parcel {
  /// See: https://developer.android.com/reference/android/os/Parcel#STRING_CREATOR
  //static const Parcelable.Creator<String> STRING_CREATOR = null; // TODO

  final Uint8Buffer _output = Uint8Buffer();
  final ByteData _buffer = ByteData(8);
  Uint8List _bufferAsList;

  Parcel.obtain() {
    _bufferAsList = _buffer.buffer.asUint8List();
  }

  void writeValue(final Object val) {
    if (val == null) {
      writeInt(VAL_NULL);
    } else if (val is String) {
      writeInt(VAL_STRING);
      writeString(val);
    } else if (val is Map<String, Object>) {
      writeInt(VAL_MAP);
      writeMap(val);
    } else if (val is Bundle) {
      // Must come before Parcelable
      writeInt(VAL_BUNDLE);
      writeBundle(val);
    } else if (val is Parcelable) {
      // Classes that implement Parcelable must come before this
      writeInt(VAL_PARCELABLE);
      writeParcelable(val, 0);
    } else if (val is int) {
      writeInt(VAL_LONG);
      writeLong(val);
    } else if (val is double) {
      writeInt(VAL_DOUBLE);
      writeDouble(val);
    } else if (val is bool) {
      writeInt(VAL_BOOLEAN);
      writeBoolean(val);
    } else if (val is Float64List) {
      // Must come before List
      writeInt(VAL_DOUBLEARRAY);
      writeDoubleArray(val);
    } else if (val is Uint8List) {
      // Must come before List
      writeInt(VAL_BYTEARRAY);
      writeByteArray(val);
    } else if (val is Int32List) {
      // Must come before List
      writeInt(VAL_INTARRAY);
      writeIntArray(val);
    } else if (val is Int64List) {
      // Must come before List
      writeInt(VAL_LONGARRAY);
      writeLongArray(val);
    } else if (val is List<bool>) {
      // Must come before List
      writeInt(VAL_BOOLEANARRAY);
      writeBooleanArray(val);
    } else if (val is List<String>) {
      // Must come before List
      writeInt(VAL_STRINGARRAY);
      writeStringArray(val);
    } else if (val is List) {
      writeInt(VAL_LIST);
      writeList(val);
    } else {
      throw ArgumentError("Parcel: unable to marshal value $val");
    }
  }

  void writeBoolean(final bool val) {
    writeInt(val ? 1 : 0);
  }

  void writeFloat(final double val) {
    _buffer.setFloat32(0, val, Endian.host);
    _write(_bufferAsList, 0, 4);
  }

  void writeDouble(final double val) {
    _buffer.setFloat64(0, val, Endian.host);
    _write(_bufferAsList, 0, 8);
  }

  void writeByte(final int val) {
    writeInt(val);
  }

  void writeInt(final int val) {
    _buffer.setInt32(0, val, Endian.host);
    _write(_bufferAsList, 0, 4);
  }

  void writeLong(final int val) {
    _buffer.setInt64(0, val, Endian.host);
    _write(_bufferAsList, 0, 8);
  }

  void writeString(final String val) {
    if (val == null) {
      return writeInt(-1);
    }
    writeInt(val.length);
    _write(Int16List.fromList(val.codeUnits) // TODO: optimize this
        .buffer
        .asUint8List());
  }

  void writeBooleanArray(final List<bool> vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    vals.forEach((val) => writeBoolean(val));
  }

  void writeFloatArray(final Float32List vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    _write(vals.buffer.asUint8List(vals.offsetInBytes, 4 * vals.length));
  }

  void writeDoubleArray(final Float64List vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    _write(vals.buffer.asUint8List(vals.offsetInBytes, 8 * vals.length));
  }

  void writeByteArray(final Uint8List vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    _write(vals);
  }

  void writeCharArray(final Int16List vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    // Can't use buffer.putInt16List() because Android uses writeInt()
    vals.forEach((val) => writeInt(val));
  }

  void writeIntArray(final Int32List vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    _write(vals.buffer.asUint8List(vals.offsetInBytes, 4 * vals.length));
  }

  void writeLongArray(final Int64List vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    _write(vals.buffer.asUint8List(vals.offsetInBytes, 8 * vals.length));
  }

  void writeStringArray(final List<String> vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    vals.forEach((val) => writeString(val));
  }

  void writeList(final List<Object> vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    vals.forEach((val) => writeValue(val));
  }

  void writeMap(final Map<String, Object> vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    vals.forEach((key, val) {
      writeValue(key);
      writeValue(val);
    });
  }

  void writeArrayMap(final Map<String, Object> vals) {
    if (vals == null) {
      return writeInt(-1);
    }
    writeInt(vals.length);
    vals.forEach((key, val) {
      writeString(key);
      writeValue(val);
    });
  }

  void writeBundle(final Bundle bundle) {
    if (bundle == null) {
      return writeInt(-1);
    }
    bundle.writeToParcel(this);
  }

  void writeParcelable(final Parcelable parcelable, [int parcelableFlags = 0]) {
    if (parcelable == null) {
      return writeString(null);
    }
    writeString(parcelable.parcelableCreator);
    parcelable.writeToParcel(this, parcelableFlags);
  }

  void _write(final Uint8List data, [int start = 0, int end]) {
    _output.addAll(data, start, end);
    if (_output.lengthInBytes % 4 != 0) {
      _writePadding(4 - _output.lengthInBytes % 4);
    }
  }

  void _writePadding(final int count) {
    for (var i = 0; i < count; i++) {
      _output.add(0);
    }
  }

  ByteData asByteData() {
    return _output.buffer.asByteData(0, _output.lengthInBytes);
  }

  Uint8List asUint8List() {
    return Uint8List.sublistView(asByteData());
  }
}
